torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -1985,6 +1985,51 @@ def test_return_local_rrefs(self):
 
         self.assertEqual(rets, [11, 12, 13])
 
+    @dist_init
+    def test_rref_type(self):
+
+        def launched_rpc(events):
+            expected_name = "rpc_sync#_rref_typeof_on_owner"
+            return any([e.name.startswith(expected_name) for e in events])
+
+        dst = worker_name((self.rank + 1) % self.world_size)
+        rref = rpc.remote(dst, torch.add, args=(torch.ones(2), 1))
+
+        with torch.autograd.profiler.profile() as p:
+            t = rref._get_type()
+
+        print(p.function_events)
+        self.assertTrue(launched_rpc(p.function_events))
+        self.assertEqual(t, type(torch.ones(2)))
+
+        with torch.autograd.profiler.profile() as p:
+            for _ in range(10):
+                t = rref._get_type()
+
+        self.assertFalse(launched_rpc(p.function_events))
+
+        rref = rpc.remote(dst, MyClass, args=(0,))
+        self.assertEqual(rref._get_type(), MyClass)
+
+    @dist_init
+    def test_rref_type_owner(self):
+        rref = RRef(torch.ones(2) + 1)
+        self.assertEqual(rref._get_type(), type(torch.ones(2)))
+
+        rref = RRef(MyClass(0))
+        self.assertEqual(rref._get_type(), MyClass)
+
+    @staticmethod
+    def _slow_add(x, y):
+        time.sleep(1)
+        return x + y
+
+    @dist_init
+    def test_rref_type_slow_init(self):
+        dst = worker_name((self.rank + 1) % self.world_size)
+        rref = rpc.remote(dst, RpcTest._slow_add, args=(torch.ones(2), 1))
```

Comment:
What happens if `rpc.remote()` call times out and then the user calls `_get_type()`? Will we just handle this error situation by the `_get_type()` RPC also timing out? Would we want to eventually error out in an "eager" way if we know `rpc.remote()` for the RRef we're trying to get the type of has already failed?

Reply:
There is `getTimedOut()` on RRefs, maybe we can use/generalize this if we want to do the above.

Reply:
It should have the same behavior as calling normal RPC on a timed-out RRef. Let me add a test to check the error message.

Reply:
Tried the a similar test as `test_rref_timeout`, it will wait until the RRef is ready, even if the `remote` function timed out. Let me create an issue to track.

Reply:
#44771